contracts/ReverseFullCheckpoint.sol

```diff
@@ -0,0 +1,427 @@
+
+            address[] memory next = HeaderReader.getEpoch(headers[x]);
+
+            // Verify subnet header meta information
+            require(validationParams.number > 0, "Repeated Genesis");
+            require(
+                validationParams.number >
+                    int256(
+                        uint256(
+                            uint64(headerTree[latestFinalizedBlock].mix >> 128)
+                        )
+                    ),
+                "Old Block"
+            );
+            require(
+                headerTree[validationParams.parentHash].mix != 0,
+                "Parent Missing"
+            );
+            require(
+                int256(
+                    uint256(
+                        uint64(
+                            headerTree[validationParams.parentHash].mix >> 128
+                        )
+                    )
+                ) +
+                    1 ==
+                    validationParams.number,
+                "Invalid N"
+            );
+            require(
+                uint64(headerTree[validationParams.parentHash].mix >> 64) <
+                    validationParams.roundNumber,
+                "Invalid RN"
+            );
+            require(
+                uint64(headerTree[validationParams.parentHash].mix >> 64) ==
+                    validationParams.prevRoundNumber,
+                "Invalid PRN"
+            );
+
+            bytes32 blockHash = keccak256(headers[x]);
+
+            // If block is the INIT_EPOCH block, prepared for validators switch
+            if (headerTree[blockHash].mix > 0) {
+                revert("Repeated Header");
+            }
+
+            // Verify subnet header certificates
+            address[] memory signerList = new address[](
+                validationParams.sigs.length
+            );
+            for (uint256 i = 0; i < validationParams.sigs.length; i++) {
+                address signer = HeaderReader.recoverSigner(
+                    validationParams.signHash,
+                    validationParams.sigs[i]
+                );
+                if (lookup[signer] != true) revert("Invalid Validator");
+                signerList[i] = signer;
+            }
+            (bool isUnique, int256 uniqueCounter) = checkUniqueness(signerList);
+            if (!isUnique) {
+                revert("Repeated Validator");
+            }
+            if (uniqueCounter * 100 < currentValidators.threshold) {
+                revert("Insufficient Signatures");
+            }
+
+            if (
+                next.length > 0 &&
+                validationParams.prevRoundNumber <
+                validationParams.roundNumber -
+                    (validationParams.roundNumber % INIT_EPOCH)
+            ) {
+                setLookup(next);
+
+                validators[validationParams.number] = Validators({
+                    set: next,
+                    threshold: int256((next.length * 2 * 100) / 3)
```

Comment:
threshold is 2/3. @liam-lai Please check that whether threshold must be 2/3, thx

Reply:
According to https://github.com/XinFinOrg/XDPoSChain/blob/master/params/config.go#L44-L46
the threshold can be changed. not necessarily *2/3.
In golang code, it is a float, such as 0.667

Reply:
fixed

Reply:
the threshold can be changed.
for example, the community decide to change threshold to 0.500 in golang code. So i suggest in smart contract, make threshold changeable too.

Reply:
take [https://github.com/XinFinOrg/XDPoSChain/blob/master/params/config.go](https://github.com/XinFinOrg/XDPoSChain/blob/master/params/config.go) as an example, see below code. The threshold change from 0.667 (default) to 0.4 (7956000 round number) to 0.667 (7974000  round number). I'm afraid in mainnet the threshold is changed in the future.
```
DevnetV2Configs = map[uint64]*V2Config{
		Default: {
			MaxMasternodes:       108,
			SwitchRound:          0,
			CertThreshold:        0.667,
			TimeoutSyncThreshold: 3,
			TimeoutPeriod:        30,
			MinePeriod:           2,
		},
		7956000: { // 2024.01.17 Devnet Deplyment Issue
			MaxMasternodes:       108,
			SwitchRound:          7956000,
			CertThreshold:        0.4,
			TimeoutSyncThreshold: 3,
			TimeoutPeriod:        30,
			MinePeriod:           2,
		},
		7974000: {
			MaxMasternodes:       108,
			SwitchRound:          7974000,
			CertThreshold:        0.667,
			TimeoutSyncThreshold: 3,
			TimeoutPeriod:        30,
			MinePeriod:           2,
		},
	}
```

Reply:
seems the problem can be solved with upgradable proxy. resolve